File: client/bindings/nodejs/examples/offline_signing/1_transaction_preparation.ts
// Copyright 2021-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Client, initLogger } from '@iota/client';
import { writeFile, readFile } from 'fs/promises';

require('dotenv').config({ path: '../../../.env' });

// From examples directory, run with:
// node ./dist/offline_signing/1_transaction_preparation.js

const ADDRESS_FILE_NAME = __dirname + '/../../offline_signing/address.json';
const PREPARED_TRANSACTION_FILE_NAME =
    __dirname + '/../../offline_signing/prepared_transaction.json';

// In this example we will get inputs and prepare a transaction
async function run() {
    initLogger();
    if (!process.env.NODE_URL) {
        throw new Error('.env NODE_URL is undefined, see .env.example');
    }
    const onlineClient = new Client({
        // Insert your node URL in the .env.
        nodes: [process.env.NODE_URL],
        localPow: true,
    });

    const address =
        'rms1qqv5avetndkxzgr3jtrswdtz5ze6mag20s0jdqvzk4fwezve8q9vkpnqlqe';
    const amount = 1000000;
    try {
        // Recovers the address from example `0_address_generation`.
        const input_address = JSON.parse(
            await readFile(ADDRESS_FILE_NAME, 'utf8'),
        );

        // Gets enough inputs related to the address to cover the amount.
        const inputs = await onlineClient.findInputs(input_address, amount);

        // Prepares the transaction
        const preparedTransaction = await onlineClient.prepareTransaction(
            undefined,
            {
                inputs,
                output: { address, amount: amount.toString() },
            },
        );

        console.log(`Prepared transaction sending ${amount} to ${address}.`);

        await writeFile(
            PREPARED_TRANSACTION_FILE_NAME,
            JSON.stringify(preparedTransaction),
        );
    } catch (error) {
        console.error(error);
    }
}

run().then(() => process.exit());

File: client/examples/offline_signing/1_transaction_preparation.rs
// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we get inputs and prepare a transaction.
//! `cargo run --example 1_transaction_preparation --release`.

use std::{
    fs::File,
    io::{prelude::*, BufWriter},
    path::Path,
};

use iota_client::{
    api::{PreparedTransactionData, PreparedTransactionDataDto},
    Client, Result,
};

const ADDRESS_FILE_NAME: &str = "examples/offline_signing/address.json";
const PREPARED_TRANSACTION_FILE_NAME: &str = "examples/offline_signing/prepared_transaction.json";

#[tokio::main]
async fn main() -> Result<()> {
    dotenv::dotenv().ok();

    let node_url = std::env::var("NODE_URL").unwrap();

    // Address to which we want to send the amount.
    let address = "rms1qruzprxum2934lr3p77t96pzlecxv8pjzvtjrzdcgh2f5exa22n6ga0vm69";
    // The amount to send.
    let amount = 1_000_000;

    // Create a client instance.
    let online_client = Client::builder()
        // Insert your node URL in the .env.
        .with_node(&node_url)?
        .finish()?;

    // Recovers addresses from example `0_address_generation`.
    let addresses = read_addresses_from_file(ADDRESS_FILE_NAME)?;
    // Gets enough inputs related to these addresses to cover the amount.
    let inputs = online_client.find_inputs(addresses, amount).await?;

    // Prepares the transaction.
    let mut transaction_builder = online_client.block();
    for input in inputs {
        transaction_builder = transaction_builder.with_input(input)?;
    }
    let prepared_transaction = transaction_builder
        .with_output(address, amount)
        .await?
        .prepare_transaction()
        .await?;

    println!("Prepared transaction sending {amount} to {address}.");

    write_prepared_transaction_to_file(PREPARED_TRANSACTION_FILE_NAME, &prepared_transaction)
}

fn read_addresses_from_file<P: AsRef<Path>>(path: P) -> Result<Vec<String>> {
    let mut file = File::open(&path).unwrap();
    let mut json = String::new();
    file.read_to_string(&mut json).unwrap();

    Ok(serde_json::from_str(&json)?)
}

fn write_prepared_transaction_to_file<P: AsRef<Path>>(
    path: P,
    prepared_transaction: &PreparedTransactionData,
) -> Result<()> {
    let json = serde_json::to_string_pretty(&PreparedTransactionDataDto::from(prepared_transaction))?;
    let mut file = BufWriter::new(File::create(path).unwrap());

    println!("{json}");

    file.write_all(json.as_bytes()).unwrap();

    Ok(())
}
